Save remove for blockwise status

boaks opened the pull request:
Add save remove for blockwise status and remove synchronization.
Add separate health loggers and improve observe tests.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

boaks commented:
All together fixes issue #429.

boaks commented:
To anticipate, that this should be split more:
- the exchange house-keeping and clean up changes the Exchange and the classes using it.
- the striped exchange execution changes the Exchange and the classes using it.
- the blockwise layer save remove changes a class using the exchange.

So the changes are pretty much interwoven and to split them would introduce the risc of errors when merging them again (and a lot of additional work).

boaks commented:
> I'm not sure to understand why we don't need synchronized for blockwiseLayer anymore, I suppose this is because we consider block transfer is relative to one exchange and will not be access by several thread thx to stripedExecutor.

The remove of the synchronization was motivated by deadlocks during tests. If you don't get these deadlock, and you get better results with that synchronization, we simply can revert and stay with the synchronized version.

boaks commented:
> Why do we need specific health logger ?

We don't need it, but from my experience, I want to check, if we still have leaks without getting "flushed" by too many other messages. Therefore I introduced the health logger.

boaks commented:
> I'm a bit disturb by the addition of the remove event on ExchangeObserver which is called to remove the exchange from the store. I feel it should behave better now than before but I found this design not so intuitive. (generally the remove event means something was removed, here its means something should be removed)

Maybe we rename the ExchangeObserver, if you fell, that is is no longer an observer.
My feeling was, that we mixed up the Exchange store house keeping by using a "complex mix of current complete/complete/set current ...". With encapsulating that complexity in clear rules in Exchange I think it's more intuitive.

boaks commented:
> About adding "complete" event on message, I'm not sure to understand "what a complete response should mean"?

It's just a auxiliary to ensure, that the last block of a blockwise notify is transmitted. I introduced to make also test, when the blockwise notifies are not mixed up.

boaks commented:
> I must confess I just understand hardly the outline.

Sure, and using the striped execution in scandium also didn't fully work with the first change, I remember two fixes (delegate timer callback back to stripe and use stripes for sending messages as well). So, yes I think, there will still be issues in the future, which must be resolved and which may change it again. But for now, it seems to be the right way.

> Using exchange as stripe is the better than before but I'm not totally convince that the "right" way.

Just work out a better way ;-).

boaks commented on californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java:
```diff
@@ -90,7 +117,14 @@
 public class Exchange {
 
+	private static final Logger LOGGER = LoggerFactory.getLogger(Exchange.class.getName());
+	private static final boolean DEBUG = true;
```
Until it really, really shows, that no issues are left, I would keep it :-)

boaks commented on californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java:
```diff
@@ -640,6 +994,107 @@ public EndpointContext getEndpointContext() {
+	/**
+	 * Assert, that the current thread owns this exchange.
+	 */
+	private void assertOwner() {
```
> other does not.

I concentrated on the setter, I fell that they are critical. Others, as setBlock1ToAck seesm for me to be not that important.

sbernard31 replied:
Those it means this is temporary stuff for debugging only ?

boaks replied:
Yes, but my intention of "temporary" may take a little longer :-).

boaks commented on californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java:
```diff
@@ -197,33 +202,33 @@ public int assignMessageId(final Message message) {
 	private int registerWithMessageId(final Exchange exchange, final Message message) {
-
 		enableStatus = true;
+		exchange.assertIncomplete(message);
```
If a Exchange is complete, it would not clean-up the store again. Therefore I assert the incompleteness (and assert the ownership with that). The point here is, that if this is violated somewhere, it's easier to assert here and see, where the violation is from, then do a analysis afterwards.

boaks replied:
Checking the owner should detect "concurrent" execution.

sbernard31 replied:
it makes sense

boaks commented:
> Another question I have in mind : if we have an endpoint identifier should it be a better stripe ? What do you feel about that ? It's not to begin a polemic but more to try to guess what could be the right design in an ideal world. (maybe this is not the good timing to talk about that)#

Just use the "endpoint context" (or what ever you want to use) and build your "endpoint identities" and try it out, if it works better to stripe on that or not.

sbernard31 commented:
> Maybe we rename the ExchangeObserver, if you fell, that is is no longer an observer.

If we integrate this PR I prefer to rename it.  But I'm not sure to know how to name it as its not a current pattern. maybe something like ExchangeContainer ? (Should it mean that ExchangeStore should implement it ?)

sbernard31 commented:
About health logging, I think it should not be activated by default maybe if should contains DEBUG or TRACE log only as is for debugging purpose.

boaks commented:
> If we integrate this PR I prefer to rename it.

RemoveHandler ?

boaks commented:
> About health logging, I think it should not be activated by default maybe if should contains DEBUG or TRACE log only as is for debugging purpose.

If we keep the separate logger, we can go for any level.

sbernard31 commented:
> If we keep the separate logger, we can go for any level.

I'm ok to keep it but I would go for max DEBUG level for it and HEALTH_STATUS_INTERVAL  to 0 by default.

boaks commented:
I renamed the `ExchangeObserver` into `RemoveHandler` and changed the levels and default of the health-status logging.
The heap/gc downgrade was mainly caused by using the `Exchange` itself as "stripe", using an other simpler identifying object improves that. But the "time" penalty of the stripe executor just stays.

To easier compare the performance with and without the striped execution, it's now possible to change
`CoapEndpoint.start()`

```
		exchangeExecutor = new StripedExecutorService(executor);
```

to 

```
		exchangeExecutor = executor;
```

sbernard31 commented:
Could you rebase this branch on 2.0.x to benefits to the fix about maven build.

Could we make the DEBUG variable in Exchange ? maybe using : 
```
static final boolean DEBUG = LOGGER.isDebugEnabled();
```

boaks commented:
> Could you rebase this branch on 2.0.x to benefits to the fix about maven build.

Sure. It would outdate all comments, therefore I didn't rebase.

boaks commented:
> Could we make the DEBUG variable in Exchange ? maybe using :

I'm not sure, if this is **NOW** a good idea :-).

boaks commented:
@sbernard31 

So, beside of the performance analysis in issue #559, is there any progress in the review?

sbernard31 commented:
As I explained, in a previous mail, I feel that will not be so worthy to go more in details about this review.

So I would not plan to go deeper.

boaks commented:
"That's why I'm currently more focus on analyzing behavior than reading the PR." 
OK, could you let us know, when your analysis of the behaviour is ready? 
And if your result is to merge this PR or not?

sbernard31 commented:
Ok my vision about this PR.

We have a thread design issue in Californium.
I would have liked a discussion about the way we should go to fix this before we start.
Anyway we could still do that later so let's see if this PR should be merged or not.

Again the size of the PR does not help to have an accurate vision of how this will behave.
Main idea is to use StripedExecutorService ensure that all changes about a given exchange is executed in same thread.
This idea is probably better than before. (even if I still have doubt that is the "right design" but we could discuss about that later in an other PR)

So if it behaves better what could be blocking issue about merging this PR.
FMPOV :
1) does #551 introduce deadlocks ? As Achim does not play so much with synchronized block and even removed some, there are not so much risks.
2) does #551 introduce larger performance regressions ? #559  shows there is regression but last changes limits this regressions. (If we really want to be confident about that we probably need more tests)
3) does  #551 introduce features regressions ? we have a lot of junit tests and it makes me confident about that.
4) does #551 make code more complicated and harder to refactor ? I don't have a bad feeling about this. (this is very subjective)

So my feeling is #551 could help at short term.
(FMPOV, If we integrate it in 2.0.x `debug` in `Exchange` should be configurable)

I would have your opinion about that too.
@boaks what is your confident level about this PR ? You should be the one which have the better vision about that.
@sophokles73 Any comments about that ?

boaks commented:
> I would have liked a discussion about the way we should go to fix this before we start.

But therefore, we have discussion on this PR.
FMPOV, it's much easier to discuss solutions, which could be tested, than theoretical ideas, which first must be implemented by someone to see, if they work.
You still free to provide an alternative approach.

boaks commented:
> what is your confident level about this PR ? You should be the one which have the better vision about that.

It's better than before, but sure not fixing all issues, especially not in blockwise.
I also guess, that it introduce some new issues, which must the be fixed afterwards, but they should be much smaller than now.

boaks commented:
See #PR #590
